### PR TITLE
[FIX][SOLAR-387] Add fallback when no registration available or FF TAO_AS_A_TOOL is disabled

### PR DESCRIPTION
**Pull request by andreluizmachado**
We should redirect the user to the default login page in case:

1. The FF is not enabled
2. The LTI Platform config is missing

Please check the [bug ticket](https://oat-sa.atlassian.net/browse/SOLAR-387) for more details.

### Current Behavior

https://github.com/oat-sa/extension-tao-lti/assets/11900046/c10f57b0-ea3c-445d-a02b-058a96e034a5

### The fix
https://github.com/oat-sa/extension-tao-lti/assets/11900046/637b7d20-3caa-4179-bbb4-dd232f85b97e

### CODE REVIEW

File: controller/AuthoringTool.php [hunk 1]
```diff
@@ -55,6 +61,9 @@ public function run(): void
         }
     }
 
+    /**
+     * @thorws LtiException
```
Comment (pribi) [inline controller/AuthoringTool.php:65]: Note: typo, throws.

**Review by pribi — APPROVED** [review]
LGTM, but not tested.

File: controller/AuthoringTool.php [hunk 2]
```diff
@@ -87,4 +99,52 @@ public function launch(): void
...
+        if (!$this->isFeatureTaoAsToolEnabled()) {
+            $this->getLogger()->info(
+                'TAO as tool feature is disabled. The user will be redirected to the login page.'
+            );
+            $this->redirect(_url('login', 'Main', 'tao'));
+        }
+
+        try {
+            $message = $this->getValidatedLtiMessagePayload();
+        } /** @noinspection PhpRedundantCatchClauseInspection */ catch (LtiException $exception) {
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:134]: Can we maybe fix this instead of ignoring it?

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
**Review Checklist**
- [ ] New code is covered by tests (if applicable)
- [ ] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [ ] New code is respecting code style rules
- [ ] New code is respecting best practices
- [ ] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [ ] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
- [x] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Commits messages are meaningful
- [x] Commits are atomic
- [ ] Changelog is updated according to changes (if applicable)
- [ ] Documentation is updated according to changes (if applicable)

File: controller/AuthoringTool.php [hunk 3]
```diff
@@ -64,13 +73,16 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
...
      * @throws InterruptedActionException
+     * @throws LtiException
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws common_exception_Error
      */
     public function launch(): void
     {
-        $message = $this->getValidatedLtiMessagePayload();
+        $message = $this->getLtiMessageOrRedirectToLogin();
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:85]:
```suggestion
        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
```

File: controller/AuthoringTool.php [hunk 4]
```diff
@@ -87,4 +99,52 @@ public function launch(): void
...
+     * @throws LtiException
+     * @throws NotFoundExceptionInterface
+     */
+    private function getLtiMessageOrRedirectToLogin(): LtiMessagePayloadInterface
+    {
+        if (!$this->isFeatureTaoAsToolEnabled()) {
+            $this->getLogger()->info(
+                'TAO as tool feature is disabled. The user will be redirected to the login page.'
+            );
+            $this->redirect(_url('login', 'Main', 'tao'));
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:129]: I would separate this business logic from that method. IMO it is violating Single Responsibility Principle

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
My recommendation is to fallback into redirect after we catches exception from . `getValidatedLtiMessagePayload` method

File: controller/AuthoringTool.php (hunk 4 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:129]: Which responsibility do you mean? It is the responsibility of a controller layer to do redirects

File: controller/AuthoringTool.php [hunk 5]
```diff
@@ -87,4 +99,52 @@ public function launch(): void
...
+        }
+
+        try {
+            $message = $this->getValidatedLtiMessagePayload();
+        } /** @noinspection PhpRedundantCatchClauseInspection */ catch (LtiException $exception) {
+            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
+                throw $exception;
+            }
+
+            $this->getLogger()->info(
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:139]: Is this an info or a warning or even error?

File: controller/AuthoringTool.php [hunk 6]
```diff
@@ -28,16 +28,22 @@
...
 use oat\taoLti\models\classes\Tool\Validation\Lti1p3Validator;
 use oat\taoLti\models\classes\user\UserService;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\NotFoundExceptionInterface;
 use tao_actions_Main;
 use tao_models_classes_UserService;
 
 class AuthoringTool extends ToolModule
 {
+    private const LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE = 'No matching registration found tool side';
```
[the remark this answers is not in the document]
Comment (andreluizmachado) [reply controller/AuthoringTool.php:45]: Do you have a suggestion?

File: controller/AuthoringTool.php (hunk 5 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:139]: Now it is an info as this flow is expected by the application

**Comment by andreluizmachado** [comment]
> My recommendation is to fallback into redirect after we catches exception from . `getValidatedLtiMessagePayload` method

If I understood correctly this is what I am doing.

File: controller/AuthoringTool.php (hunk 4 above)
Comment (bartlomiejmarszal) [reply controller/AuthoringTool.php:129]: it either geting the LtiMessage or it redirect to login

File: controller/AuthoringTool.php (hunk 5 above)
Comment (bartlomiejmarszal) [reply controller/AuthoringTool.php:139]: Is it expected to throw an error when trying to login into authoring tool?

File: controller/AuthoringTool.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (bartlomiejmarszal) [reply controller/AuthoringTool.php:45]: SOrry I just understood purpose of it. I do not recomend catching particular errors by their messaage content. WHat if somebody add some new string?

File: controller/AuthoringTool.php (hunk 4 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:129]: If you are talking about the method itself then, in this case, it loses its proposes so I can just move this logic back to the controller. Honestly, I do not like the idea as the method explains exactly what he is supposed to do. Asisde from that the Single Responsability principle applies to the class itself, not methods.

File: controller/AuthoringTool.php (hunk 5 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:139]: We do not throw an error, but we throw an **exception** (LTIException).

Comment (andreluizmachado) [reply controller/AuthoringTool.php:139]: yes, in this case, it is expected someone trying to access it without the proper config. in this case what we want to do it to redirect the user to the normal login page.

File: controller/AuthoringTool.php [hunk 7]
```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
...
      * @throws InterruptedActionException
+     * @throws LtiException
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws common_exception_Error
      */
     public function launch(): void
     {
-        $message = $this->getValidatedLtiMessagePayload();
+        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:85]:
```suggestion
        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
        try {
            $message = $this->getValidatedLtiMessagePayload();
        } catch (Exception $e) {
            $this->logger->warning;
            $this-redirect();
        }
```

File: controller/AuthoringTool.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (andreluizmachado) [reply controller/AuthoringTool.php:45]: ok, but this is not possible @bartlomiejmarszal

File: controller/AuthoringTool.php [hunk 8]
```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
...
+            $this->getLogger()->info(
+                'TAO as tool feature is disabled. The user will be redirected to the login page.'
+            );
+            $this->redirect(_url('login', 'Main', 'tao'));
+        }
+
+        try {
+            $message = $this->getValidatedLtiMessagePayload();
+        } catch (LtiException $exception) {
+            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:135]: We should use Exception code or Exception type to define this not message string.

Comment (Karol-Stelmaczonek) [reply controller/AuthoringTool.php:135]: Sometimes exceptions don't implement the code and its hard to add it, maybe that's the case. But of course better would be to use code of specific exception class to check.

**Review by bartlomiejmarszal — CHANGES_REQUESTED** [review]
IMO this is the fallback mechanism and we could title this PR as `add fallback when no registration available`. This is an important change and we should be careful with our changes and we should describe them with clarity. Also, I would say it is a feature, not a bug.

File: controller/AuthoringTool.php [hunk 9]
```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
...
             ->getContainer()
             ->get(LtiService::class)
-            ->startLti1p3Session($message, $user);
+            ->startLti1p3Session($ltiMessage, $user);
 
         $this->forward('run', null, null, $_GET);
     }
+
+    /**
+     * @return bool
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:104]:
```suggestion
```

File: controller/AuthoringTool.php [hunk 10]
```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
...
+    private function isFeatureTaoAsToolEnabled(): bool
+    {
+        return $this->getServiceManager()
+            ->getContainer()
+            ->get(FeatureFlagChecker::class)
+            ->isEnabled(FeatureFlagCheckerInterface::FEATURE_FLAG_TAO_AS_A_TOOL);
+    }
+
+    /**
+     * @return LtiMessagePayloadInterface
```
Comment (bartlomiejmarszal) [inline controller/AuthoringTool.php:117]:
```suggestion
```

File: controller/AuthoringTool.php (hunk 4 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:129]: If I had 2 redirects like in your example, I would just rename it to `getLtiMessageOrPerformRedirect` that would be the simplest way.

File: controller/AuthoringTool.php (hunk 5 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:139]: but this is what is being defined by the product and we should reflect what the product says in our code base. So you guys are saying that is not an expected behavior? Can you guys point me out why? Based on what?

Comment (andreluizmachado) [reply controller/AuthoringTool.php:139]: I would go with a warning on this, it is fine for me, but I do not think it is an error.

File: controller/AuthoringTool.php (hunk 9 above)
Comment (andreluizmachado) [reply controller/AuthoringTool.php:104]: I am removing it, but just to let my opinion here. I do not see the gain of this, especially using IDEs to autocomplete it, I will have to care about customizing my IDE to change its behavior of remove it by myself.

**Comment by andreluizmachado** [comment]
> IMO this is the fallback mechanism and we could title this PR as `add fallback when no registration available`. This is an important change and we should be careful with our changes and we should describe them with clarity. Also, I would say it is a feature, not a bug.

It's not common to refer to a "fallback" as a "feature." Typically, a "fallback" is a contingency strategy or action triggered when something doesn't work as expected, or when a system cannot perform its primary task. For example, in programming, a "fallback" can be an alternative function that is executed if the main function fails. In security systems, a "fallback" can be a backup procedure in case the primary systems fail.

A "feature" is usually a planned and desired functionality of a system, product, or service that is designed to meet specific user needs. Features are intentional components of a system, while a "fallback" is a contingency response to unforeseen situations.

In summary, it's not appropriate to say that a "fallback" is a "feature" because they serve different roles in a system or product.

File: controller/AuthoringTool.php (hunk 8 above)
Comment (bartlomiejmarszal) [reply controller/AuthoringTool.php:133]: Can you make changes to the exception so it will be easy to recognise the type of exception?

Comment (andreluizmachado) [reply controller/AuthoringTool.php:133]: I would avoid creating a new release on Lti1p3Core only for this.

**Review by Karol-Stelmaczonek — APPROVED** [review]
